From 7c6bde8536c594763ad850873041b2c3d4bf758d Mon Sep 17 00:00:00 2001
From: Francesco VIRLINZI <francesco.virlinzi@st.com>
Date: Tue, 3 Mar 2009 14:49:06 +0100
Subject: [PATCH] stm_lirc: added STLIRC_DEVICE macro and debug support via Kconfigurator

This patch adds
 - the STLIRC_DEVICE macro in the setup-xxxx.c files
 - the wakeup interrupt support in lirc device driver
 - the printk debug support via Kconfigurator

Signed-off-by: Francesco Virlinzi <francesco.virlinzi@st.com>
---
 arch/sh/kernel/cpu/sh4/setup-stb7100.c |   25 ++-----------------------
 arch/sh/kernel/cpu/sh4/setup-stx5197.c |   29 ++++++-----------------------
 arch/sh/kernel/cpu/sh4/setup-stx7105.c |   25 ++-----------------------
 arch/sh/kernel/cpu/sh4/setup-stx7111.c |   25 ++-----------------------
 arch/sh/kernel/cpu/sh4/setup-stx7141.c |   25 ++-----------------------
 arch/sh/kernel/cpu/sh4/setup-stx7200.c |   25 ++-----------------------
 drivers/char/lirc/Kconfig              |   11 +++++++++--
 drivers/char/lirc/lirc_stm.c           |   17 +++++++++++++++--
 include/linux/stm/soc_init.h           |   23 +++++++++++++++++++++++
 9 files changed, 63 insertions(+), 142 deletions(-)

diff --git a/arch/sh/kernel/cpu/sh4/setup-stb7100.c b/arch/sh/kernel/cpu/sh4/setup-stb7100.c
index 6c1b2b4..7ea814f 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stb7100.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stb7100.c
@@ -640,29 +640,8 @@ static struct plat_lirc_data lirc_private_info = {
 #endif
 };
 
-static struct resource lirc_resource[]= {
-        [0] = {
-		.start = 0x18018000,
-		.end   = 0x18018000 + 0xa0,
-	        .flags = IORESOURCE_MEM
-	},
-	[1] = {
-	        .start = 125,
-		.end   = 125,
-	        .flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device lirc_device = {
-	.name           = "lirc",
-	.id             = -1,
-	.num_resources  = ARRAY_SIZE(lirc_resource),
-	.resource       = lirc_resource,
-	.dev = {
-		   .power.can_wakeup = 1,
-	           .platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device =
+	STLIRC_DEVICE(0x18018000, 125, NULL);
 
 void __init stx7100_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/arch/sh/kernel/cpu/sh4/setup-stx5197.c b/arch/sh/kernel/cpu/sh4/setup-stx5197.c
index 1fa93a1..1b58c1c 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx5197.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx5197.c
@@ -489,31 +489,14 @@ static struct plat_lirc_data lirc_private_info = {
 	.sysclkdiv	= 1,
 	.rxpolarity	= 1,
 	.pio_pin_arr  = lirc_pios,
-	.num_pio_pins = ARRAY_SIZE(lirc_pios)
-};
-
-static struct resource lirc_resource[] = {
-	[0] = {
-		.start = 0xfd118000,
-		.end   = 0xfd118000 + 0xa8,
-		.flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = ILC_IRQ(19),
-		.end   = ILC_IRQ(19),
-		.flags = IORESOURCE_IRQ
-	},
+	.num_pio_pins = ARRAY_SIZE(lirc_pios),
+#ifdef CONFIG_PM
+	.clk_on_low_power = XTAL,
+#endif
 };
 
-static struct platform_device lirc_device = {
-	.name           = "lirc",
-	.id             = -1,
-	.num_resources  = ARRAY_SIZE(lirc_resource),
-	.resource       = lirc_resource,
-	.dev = {
-		.platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device = 
+	STLIRC_DEVICE(0xfd118000, ILC_IRQ(19), ILC_EXT_IRQ(4);
 
 void __init stx5197_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7105.c b/arch/sh/kernel/cpu/sh4/setup-stx7105.c
index 78eba9f..2127b4f 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7105.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7105.c
@@ -942,29 +942,8 @@ static struct plat_lirc_data lirc_private_info = {
 #endif
 };
 
-static struct resource lirc_resource[]= {
-        [0] = {
-		.start = 0xfd018000,
-		.end   = 0xfd018000 + 0xa0,
-	        .flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = evt2irq(0x11a0),
-		.end   = evt2irq(0x11a0),
-	        .flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device lirc_device = {
-	.name           = "lirc",
-	.id             = -1,
-	.num_resources  = ARRAY_SIZE(lirc_resource),
-	.resource       = lirc_resource,
-	.dev = {
-		   .power.can_wakeup = 1,
-	           .platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device =
+	STLIRC_DEVICE(0xfd018000, evt2irq(0x11a0), ILC_EXT_IRQ(4));
 
 void __init stx7105_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7111.c b/arch/sh/kernel/cpu/sh4/setup-stx7111.c
index dec8f0c..e18d9cb 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7111.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7111.c
@@ -575,29 +575,8 @@ static struct plat_lirc_data lirc_private_info = {
 #endif
 };
 
-static struct resource lirc_resource[]= {
-        [0] = {
-		.start = 0xfd018000,
-		.end   = 0xfd018000 + 0xa0,
-	        .flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = evt2irq(0x11a0),
-		.end   = evt2irq(0x11a0),
-	        .flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device lirc_device = {
-	.name           = "lirc",
-	.id             = -1,
-	.num_resources  = ARRAY_SIZE(lirc_resource),
-	.resource       = lirc_resource,
-	.dev = {
-		   .power.can_wakeup = 1,
-	           .platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device =
+	STLIRC_DEVICE(0xfd018000, evt2irq(0x11a0), ILC_EXT_IRQ(4));
 
 void __init stx7111_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7141.c b/arch/sh/kernel/cpu/sh4/setup-stx7141.c
index 47ab8eb..659642b 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7141.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7141.c
@@ -1038,29 +1038,8 @@ static struct plat_lirc_data lirc_private_info = {
 #endif
 };
 
-static struct resource lirc_resource[] = {
-	[0] = {
-		.start = 0xfd018000,
-		.end   = 0xfd018000 + 0xa0,
-		.flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = ILC_IRQ(81),
-		.end   = ILC_IRQ(81),
-		.flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device lirc_device = {
-	.name		= "lirc",
-	.id		= -1,
-	.num_resources	= ARRAY_SIZE(lirc_resource),
-	.resource	= lirc_resource,
-	.dev = {
-		   .power.can_wakeup = 1,
-		   .platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device =
+	STLIRC_DEVICE(0xfd018000, ILC_IRQ(81), ILC_IRQ(86));
 
 void __init stx7141_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/arch/sh/kernel/cpu/sh4/setup-stx7200.c b/arch/sh/kernel/cpu/sh4/setup-stx7200.c
index 44426e1..15485bb 100644
--- a/arch/sh/kernel/cpu/sh4/setup-stx7200.c
+++ b/arch/sh/kernel/cpu/sh4/setup-stx7200.c
@@ -888,29 +888,8 @@ static struct plat_lirc_data lirc_private_info = {
 #endif
 };
 
-static struct resource lirc_resource[]= {
-        [0] = {
-		.start = 0xfd018000,
-		.end   = 0xfd018000 + 0xa0,
-	        .flags = IORESOURCE_MEM
-	},
-	[1] = {
-		.start = ILC_IRQ(116),
-		.end   = ILC_IRQ(116),
-	        .flags = IORESOURCE_IRQ
-	},
-};
-
-static struct platform_device lirc_device = {
-	.name           = "lirc",
-	.id             = -1,
-	.num_resources  = ARRAY_SIZE(lirc_resource),
-	.resource       = lirc_resource,
-	.dev = {
-		   .power.can_wakeup = 0x1,
-	           .platform_data = &lirc_private_info
-	}
-};
+static struct platform_device lirc_device =
+	STLIRC_DEVICE(0xfd018000, ILC_IRQ(116), ILC_IRQ(117));
 
 void __init stx7200_configure_lirc(lirc_scd_t *scd)
 {
diff --git a/drivers/char/lirc/Kconfig b/drivers/char/lirc/Kconfig
index f1152df..4b52ffa 100644
--- a/drivers/char/lirc/Kconfig
+++ b/drivers/char/lirc/Kconfig
@@ -261,7 +261,7 @@ config LIRC_SUPPORT
 			depends on LIRC_SIR
         config LIRC_STM
                 tristate "STM Lirc support"
-                depends on LIRC_SUPPORT && (SH_ST_MB411 || SH_ST_MB442 || SH_ST_MB448 || SH_ST_HMS1 || SH_ST_MB520 || SH_ST_MB618 || SH_ST_MB680 || SH_ST_MB704 || SH_ST_HMP7100 || SH_ST_MB839)
+                depends on LIRC_SUPPORT && CPU_SUBTYPE_ST40
 		help
 		  The IR and UHF are identical and independent, except
 		  that the IRB receiver does not use the noise filter.
@@ -270,9 +270,16 @@ config LIRC_SUPPORT
 
         config LIRC_STM_UHF_RX
 	        bool "UHF Receiver mode"
-		depends on LIRC_STM && (SH_ST_MB411 || SH_ST_MB442 || SH_ST_MB520 || SH_ST_MB618 || SH_ST_MB680 || SH_ST_MB704 || SH_ST_HMP7100 || SH_ST_MB839)
+		depends on LIRC_STM
 		help
 		  This causes the driver to use UHF RX mode.
 
+ 	config LIRC_STM_DEBUG
+ 		depends on LIRC_STM
+ 		bool "Lirc debug"
+ 		default n
+ 		help
+ 		  Say yes if you want enable some debug printk information
+
 endmenu
 
diff --git a/drivers/char/lirc/lirc_stm.c b/drivers/char/lirc/lirc_stm.c
index 3b17d9c..8e2910c 100644
--- a/drivers/char/lirc/lirc_stm.c
+++ b/drivers/char/lirc/lirc_stm.c
@@ -76,8 +76,7 @@
 #define LIRC_STM_NAME "lirc_stm"
 
 /* General debugging */
-#undef LIRC_STM_DEBUG
-#ifdef  LIRC_STM_DEBUG
+#ifdef CONFIG_LIRC_STM_DEBUG
 #define DPRINTK(fmt, args...) printk(KERN_INFO LIRC_STM_NAME ": %s: " fmt, __FUNCTION__ , ## args)
 #else
 #define DPRINTK(fmt, args...)
@@ -1174,6 +1173,20 @@ static int lirc_stm_probe(struct platform_device *pdev)
 		return -EIO;
 	}
 
+	/* Enable wakeup interrupt if any */
+	if ((irb_irq = platform_get_irq(pdev, 1)) == 0) {
+		printk(KERN_WARNING LIRC_STM_NAME
+			": Wake IRQ configuration not found\n");
+	} else {
+		DPRINTK("IRB irq is %d\n", irb_irq);
+		if (devm_request_irq(dev, irb_irq, lirc_stm_interrupt, IRQF_DISABLED,
+					LIRC_STM_NAME, (void *)&pd) < 0) {
+			printk(KERN_ERR LIRC_STM_NAME ": IRQ register failed\n");
+			return -EIO;
+			}
+		disable_irq(irb_irq);
+		enable_irq_wake(irb_irq);
+		}
 	/* Configure for ir or uhf. uhf_switch==1 is UHF */
 	if (uhf_switch)
 		ir_or_uhf_offset = 0x40;
diff --git a/include/linux/stm/soc_init.h b/include/linux/stm/soc_init.h
index 4bfdb02..ac7c865 100644
--- a/include/linux/stm/soc_init.h
+++ b/include/linux/stm/soc_init.h
@@ -1,3 +1,26 @@
+#define STLIRC_DEVICE(_mem_start, _irq, _wake_irq)	\
+{							\
+	.name = "lirc",					\
+	.id = -1,					\
+	.num_resources = 3,				\
+	.resource = (struct resource[]) {		\
+		{					\
+			.start = _mem_start,		\
+			.end = _mem_start + 0xa0,	\
+			.flags = IORESOURCE_MEM,	\
+		}, {					\
+			.start = _irq,			\
+			.flags = IORESOURCE_IRQ,	\
+		}, {					\
+			.start = _wake_irq,		\
+			.flags = IORESOURCE_IRQ,	\
+		},					\
+	},						\
+	.dev = {					\
+		.power.can_wakeup = 1,			\
+		.platform_data = &lirc_private_info	\
+        },						\
+}
 
 #define STPIO_DEVICE(_id, _base, _irq)					\
 {									\
-- 
1.6.0.6

